File: quolance-api/src/main/java/com/quolance/quolance_api/controllers/AuthController.java
package com.quolance.quolance_api.controllers;

import com.quolance.quolance_api.dtos.users.LoginRequestDto;
import com.quolance.quolance_api.dtos.users.UserResponseDto;
import com.quolance.quolance_api.dtos.users.VerifyEmailDto;
import com.quolance.quolance_api.entities.User;
import com.quolance.quolance_api.services.auth.AuthService;
import com.quolance.quolance_api.services.entity_services.UserService;
import com.quolance.quolance_api.util.SecurityUtil;
import io.swagger.v3.oas.annotations.Operation;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/auth")
@Slf4j
public class AuthController {

    private final AuthService authService;
    private final UserService userService;


    @PostMapping("/login")
    public ResponseEntity<Void> login(
            HttpServletRequest request, HttpServletResponse response, @Valid @RequestBody LoginRequestDto body) {
        authService.login(request, response, body);
        User user = SecurityUtil.getAuthenticatedUser();
        log.info("User with ID {} logged in", user.getId());
        return ResponseEntity.ok().build();
    }

    @GetMapping("/me")
    public ResponseEntity<UserResponseDto> getSession(HttpServletRequest request) {
        return ResponseEntity.ok(authService.getSession(request));
    }

    @PostMapping("/logout")
    public ResponseEntity<Void> logout(HttpServletRequest request, HttpServletResponse response) {
        User user = SecurityUtil.getAuthenticatedUser();
        authService.logout(request, response);
        log.info("User with ID {} logged out", user.getId());
        return ResponseEntity.ok().build();
    }

    /**
     * We don't have to do anything in this endpoint, the CsrfFilter will handle it.
     * This endpoint should be invoked by the frontend to get the CSRF token.
     */
    @GetMapping("/csrf")
    public ResponseEntity<Void> csrf() {
        return ResponseEntity.ok().build();
    }

    @PostMapping("/verify-email/{email}")
    @Operation(summary = "Verify the email of the user")
    public ResponseEntity<String> verifyEmail(@PathVariable String email, @RequestBody VerifyEmailDto verificationDto) {
        String verificationResponse = userService.verifyEmail(email, verificationDto);
        return ResponseEntity.ok(verificationResponse);
    }

    @PostMapping("/resend-verification/{email}")
    @Operation(
            summary = "Resend verification code"
    )
    public ResponseEntity<String> resendVerificationEmail(@PathVariable String email) {
        userService.resendVerificationEmail(email);
        return ResponseEntity.ok("Code sent successfully");
    }
}

File: quolance-scripts/end2end-test/controllers/auth_controller.py
from base_request import BaseRequest

class AuthController(BaseRequest):
    def __init__(self, base_url):
        super().__init__(base_url)

    def login(self, login_data):
        """
        Sends a POST request to /api/auth/login to log in a user.
        """
        headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }
        response = self.post("/api/auth/login", json=login_data, headers=headers)
        return response

    def get_session(self):
        """
        Sends a GET request to /api/auth/me to retrieve the current user's session.
        """
        return self.get("/api/auth/me")

    def logout(self):
        """
        Sends a POST request to /api/auth/logout to log out the current user.
        """
        return self.post("/api/auth/logout")

    def get_csrf_token(self):
        """
        Sends a GET request to /api/auth/csrf to retrieve a CSRF token.
        """
        return self.get("/api/auth/csrf")
